CRM/Contribute/BAO/Contribution.php

```diff
@@ -5410,95 +5411,84 @@ protected static function isPaymentInstrumentChange(&$params, $pendingStatuses)
    * @throws \CiviCRM_API3_Exception
    */
   protected static function updateMembershipBasedOnCompletionOfContribution($contribution, $primaryContributionID, $changeDate) {
-    $contribution->loadRelatedMembershipObjects();
-    if (empty($contribution->_relatedObjects['membership'])) {
+    // Load memberships from contribution info. There may be multiple memberships (or none)
+    $membershipFields = array('id', 'contact_id', 'is_test', 'membership_type_id', 'contribution_recur_id', 'status_id.is_current_member',
+      'start_date', 'join_date', 'end_date', 'status_id');
+    $memberships = array();
+    if (!empty($contribution->contribution_recur_id)) {
+      // Load memberships associated with recurring contribution
+      $membershipResult = civicrm_api3('Membership', 'get', array(
+        'contribution_recur_id' => $contribution->contribution_recur_id,
+        'return' => $membershipFields,
+        'options' => ['limit' => 0],
+      ));
+    }
+    elseif (!empty($primaryContributionID)) {
+      // Load memberships associated with original contribution (linked via a MembershipPayment record)
+      $membershipPaymentResult = civicrm_api3('MembershipPayment', 'get', array(
+        'contribution_id' => $primaryContributionID,
+        'options' => ['limit' => 0],
+      ));
+      if (!empty($membershipPaymentResult['count'])) {
+        foreach ($membershipPaymentResult['values'] as $payment) {
+          $membershipIDs[] = $payment['membership_id'];
```

Comment:
@mattwire just thinking should there be some checking to try to just have unique membership_ids?

Reply:
There is no need in this case, as the result is used for the Membership.get API id => IN - it doesn't matter if duplicate IDs are passed in as they end up in a where clause and only result in a single membership record per ID.